# Ford GoBike System Data Flights Exploration
## by *Furawa*

## Data Wrangling

The Ford GoBike System Data Flights includes information about individual rides made in a bike-sharing system covering the greater San Francisco Bay area.  We will use the datasets of all the year 2018 which are divided in 12 separated datasets(1 dataset for each month of the year). 
We will download all the 12 datasets for the year 2018, gather them all together in one unique dataset, assess it and clean it if necessary. 
First of all let us import libraries useful for the process.  

In [2]:
# Import libraries
import numpy as np
import pandas as pd
import seaborn as sb
import matplotlib.pyplot as plt  
import requests
import os
import zipfile
import re
import datetime

%matplotlib inline

Let us retrieve all the urls from the [Bay Wheels trip history data](https://s3.amazonaws.com/baywheels-data/index.html).

In [4]:
# Create an empty list to store the urls
baywheels_urls = []
# Create a for loop to retrieve the urls one by one, there are 12
for i in range(1,13):
    # Remove the 0 (after 2018) after the 9th file
    if i < 10:
        url = 'https://s3.amazonaws.com/baywheels-data/20180' + str(i) +'-fordgobike-tripdata.csv.zip'
    else:
        url = 'https://s3.amazonaws.com/baywheels-data/2018' + str(i) +'-fordgobike-tripdata.csv.zip'
    baywheels_urls.append(url)
baywheels_urls

['https://s3.amazonaws.com/baywheels-data/201801-fordgobike-tripdata.csv.zip',
 'https://s3.amazonaws.com/baywheels-data/201802-fordgobike-tripdata.csv.zip',
 'https://s3.amazonaws.com/baywheels-data/201803-fordgobike-tripdata.csv.zip',
 'https://s3.amazonaws.com/baywheels-data/201804-fordgobike-tripdata.csv.zip',
 'https://s3.amazonaws.com/baywheels-data/201805-fordgobike-tripdata.csv.zip',
 'https://s3.amazonaws.com/baywheels-data/201806-fordgobike-tripdata.csv.zip',
 'https://s3.amazonaws.com/baywheels-data/201807-fordgobike-tripdata.csv.zip',
 'https://s3.amazonaws.com/baywheels-data/201808-fordgobike-tripdata.csv.zip',
 'https://s3.amazonaws.com/baywheels-data/201809-fordgobike-tripdata.csv.zip',
 'https://s3.amazonaws.com/baywheels-data/201810-fordgobike-tripdata.csv.zip',
 'https://s3.amazonaws.com/baywheels-data/201811-fordgobike-tripdata.csv.zip',
 'https://s3.amazonaws.com/baywheels-data/201812-fordgobike-tripdata.csv.zip']

Now we download all the zip files from the urls and store them in the created folder.  

In [5]:
zip_folder = 'baywheels_monthly_data_zip'   # folder to store the zip files
# Create zip_folder if it does not exist
if not os.path.exists(zip_folder):
    os.makedirs(zip_folder)
    
# Download each zip file from the urls with a for loop
for url in baywheels_urls:  
    response = requests.get(url)
    # Retrieve programmatically the name of the file which is the last part after /
    with open(os.path.join(zip_folder, url.split('/')[-1]), mode = 'wb') as file:
        file.write(response.content)
sorted(os.listdir(zip_folder))

['201801-fordgobike-tripdata.csv.zip',
 '201802-fordgobike-tripdata.csv.zip',
 '201803-fordgobike-tripdata.csv.zip',
 '201804-fordgobike-tripdata.csv.zip',
 '201805-fordgobike-tripdata.csv.zip',
 '201806-fordgobike-tripdata.csv.zip',
 '201807-fordgobike-tripdata.csv.zip',
 '201808-fordgobike-tripdata.csv.zip',
 '201809-fordgobike-tripdata.csv.zip',
 '201810-fordgobike-tripdata.csv.zip',
 '201811-fordgobike-tripdata.csv.zip',
 '201812-fordgobike-tripdata.csv.zip']

All the zip files are in the folder, we can proceed and unzip them.  

In [6]:
csv_folder = 'baywheels_monthly_data_csv'   # folder to store the unzip csv files
# Create csv_folder
if not os.path.exists(csv_folder):
    os.makedirs(csv_folder)
    
import glob
files = glob.glob(zip_folder+'/*.zip')
for file in files:
    with zipfile.ZipFile(file, 'r') as my_zip:
        my_zip.extractall(csv_folder)

In [7]:
sorted(os.listdir(csv_folder))

['201801-fordgobike-tripdata.csv',
 '201802-fordgobike-tripdata.csv',
 '201803-fordgobike-tripdata.csv',
 '201804-fordgobike-tripdata.csv',
 '201805-fordgobike-tripdata.csv',
 '201806-fordgobike-tripdata.csv',
 '201807-fordgobike-tripdata.csv',
 '201808-fordgobike-tripdata.csv',
 '201809-fordgobike-tripdata.csv',
 '201810-fordgobike-tripdata.csv',
 '201811-fordgobike-tripdata.csv',
 '201812-fordgobike-tripdata.csv']

In [84]:
# Read one file to see if everything is ok
test = pd.read_csv('baywheels_monthly_data_csv/201801-fordgobike-tripdata.csv')
test.shape
print(test.dtypes)
test.sample(6)

duration_sec                 int64
start_time                  object
end_time                    object
start_station_id             int64
start_station_name          object
start_station_latitude     float64
start_station_longitude    float64
end_station_id               int64
end_station_name            object
end_station_latitude       float64
end_station_longitude      float64
bike_id                      int64
user_type                   object
member_birth_year          float64
member_gender               object
bike_share_for_all_trip     object
dtype: object


duration_sec                start_time                  end_time  \
704             372  2018-01-31 18:52:46.6920  2018-01-31 18:58:58.9290   
83640           318  2018-01-05 08:26:04.1060  2018-01-05 08:31:22.9710   
65742           860  2018-01-12 09:13:59.7090  2018-01-12 09:28:20.6200   
56210           951  2018-01-16 08:20:26.4320  2018-01-16 08:36:18.1320   
42127           759  2018-01-19 18:45:07.1350  2018-01-19 18:57:46.3310   
60818          1045  2018-01-14 00:01:12.0370  2018-01-14 00:18:37.5650   

       start_station_id                              start_station_name  \
704                  37                             2nd St at Folsom St   
83640                27                         Beale St at Harrison St   
65742                30  San Francisco Caltrain (Townsend St at 4th St)   
56210                84                                     Duboce Park   
42127               107                           17th St at Dolores St   
60818               105                           16th St at Prosper St   

       start_station_latitude  start_station_longitude  end_station_id  \
704                 37.785000              -122.395936              64   
83640               37.788059              -122.391865              49   
65742               37.776598              -122.395282               9   
56210               37.769201              -122.433812              19   
42127               37.763015              -122.426497             139   
60818               37.764285              -122.431804             133   

                               end_station_name  end_station_latitude  \
704                        5th St at Brannan St             37.776754   
83640                       S Park St at 3rd St             37.780760   
65742                    Broadway at Battery St             37.798572   
56210                      Post St at Kearny St             37.788975   
42127  Garfield Square (25th St at Harrison St)             37.751017   
60818                    Valencia St at 22nd St             37.755213   

       end_station_longitude  bike_id   user_type  member_birth_year  \
704              -122.399018      762  Subscriber             1974.0   
83640            -122.394989     1570  Subscriber             1992.0   
65742            -122.400869     1106  Subscriber             1958.0   
56210            -122.403452     1211  Subscriber             1986.0   
42127            -122.411901     2273    Customer                NaN   
60818            -122.420975      488  Subscriber                NaN   

      member_gender bike_share_for_all_trip  
704            Male                      No  
83640          Male                      No  
65742          Male                      No  
56210        Female                      No  
42127           NaN                      No  
60818           NaN                      No

Now that we have all the files we can read them and join them together in a unique dataframe.

In [15]:
# Select all the files in the csv_folder and store them in a list
list_files = glob.glob(csv_folder+'/*.csv')
# Read all the files of the list and put them all in one unique file
baywheels_data = pd.concat(map(pd.read_csv, list_files))

In [19]:
print(baywheels_data.shape)  
print(baywheels_data.dtypes)
baywheels_data.head()

(1863721, 16)
duration_sec                 int64
start_time                  object
end_time                    object
start_station_id           float64
start_station_name          object
start_station_latitude     float64
start_station_longitude    float64
end_station_id             float64
end_station_name            object
end_station_latitude       float64
end_station_longitude      float64
bike_id                      int64
user_type                   object
member_birth_year          float64
member_gender               object
bike_share_for_all_trip     object
dtype: object


duration_sec                start_time                  end_time  \
0         71766  2018-03-31 16:58:33.1490  2018-04-01 12:54:39.2630   
1         62569  2018-03-31 19:03:35.9160  2018-04-01 12:26:25.0350   
2         56221  2018-03-31 20:13:13.5640  2018-04-01 11:50:14.8400   
3         85844  2018-03-31 11:28:07.6580  2018-04-01 11:18:52.6130   
4          1566  2018-03-31 23:37:56.6400  2018-04-01 00:04:02.8930   

   start_station_id            start_station_name  start_station_latitude  \
0               4.0   Cyril Magnin St at Ellis St               37.785881   
1              78.0           Folsom St at 9th St               37.773717   
2             258.0   University Ave at Oxford St               37.872355   
3             186.0        Lakeside Dr at 14th St               37.801319   
4             193.0  Grand Ave at Santa Clara Ave               37.812744   

   start_station_longitude  end_station_id               end_station_name  \
0              -122.408915             6.0  The Embarcadero at Sansome St   
1              -122.411647            47.0          4th St at Harrison St   
2              -122.266447           239.0  Bancroft Way at Telegraph Ave   
3              -122.262642           340.0        Harmon St at Adeline St   
4              -122.247215           196.0        Grand Ave at Perkins St   

   end_station_latitude  end_station_longitude  bike_id   user_type  \
0             37.804770            -122.403234      341    Customer   
1             37.780955            -122.399749      536  Subscriber   
2             37.868813            -122.258764     3245    Customer   
3             37.849735            -122.270582     3722    Customer   
4             37.808894            -122.256460     2355  Subscriber   

   member_birth_year member_gender bike_share_for_all_trip  
0             1964.0        Female                      No  
1             1984.0          Male                      No  
2             1983.0          Male                      No  
3                NaN           NaN                      No  
4             1979.0          Male                      No

All the files are in one unique dataframe `baywheels_data`, but there are many issues to fix before using the 
dataframe. Many columns does not have the correct data type(start_time, end_time, start_station_id,end_station_id,
user_type,member_birth_year,member_gender).

In [28]:
# Change the data type of start_time and end_time from object to date_time
baywheels_data.start_time = pd.to_datetime(baywheels_data.start_time, format = '%Y-%m-%d %H:%M:%S.%f')
baywheels_data.end_time = pd.to_datetime(baywheels_data.end_time, format = '%Y-%m-%d %H:%M:%S.%f')

In [31]:
# Assert that the changes are correct, no output means it is correct
assert baywheels_data.start_time.dtypes == '<datetime64[ns]'
assert baywheels_data.end_time.dtypes == '<datetime64[ns]'

In [49]:
# Change member_birth_year from float to int 
baywheels_data.member_birth_year = baywheels_data.member_birth_year.fillna(0) # fill all the NaN with 0 
baywheels_data.member_birth_year = baywheels_data.member_birth_year.astype(int)

In [52]:
# Check if the data type is correct, no output means it is correct 
assert baywheels_data.member_birth_year.dtypes == 'int64'

In [63]:
# Change the data type of member_gender from string to category
baywheels_data.member_gender = baywheels_data.member_gender.fillna('Other') # Change the na values with other
baywheels_data.member_gender = baywheels_data.member_gender.astype('category')

In [64]:
# Check if the data type is correct, no output means it is correct
assert baywheels_data.member_gender.dtypes == 'category'

In [66]:
# Change the user type from string to category  
baywheels_data.user_type = baywheels_data.user_type.astype('category')

In [67]:
# Check the changes, no output means it is correct
assert baywheels_data.user_type.dtypes == 'category'

In [71]:
baywheels_data.start_station_id = baywheels_data.start_station_id.fillna(0) # Replace the NaNs values with 0 
baywheels_data.end_station_id = baywheels_data.end_station_id.fillna(0)     # Replace the NaNs values with 0 
# Change data type of start_station_id and end_station_id from float to int
baywheels_data.start_station_id = baywheels_data.start_station_id.astype(int)
baywheels_data.end_station_id = baywheels_data.end_station_id.astype(int)

In [75]:
# Check the changes, no output means it is ok
assert baywheels_data.start_station_id.dtypes == 'int64'
assert baywheels_data.end_station_id.dtypes == 'int64'

In [82]:
# Some descriptives statistics for numeric variables
baywheels_data.describe()
baywheels_data.columns

Index(['duration_sec', 'start_time', 'end_time', 'start_station_id',
       'start_station_name', 'start_station_latitude',
       'start_station_longitude', 'end_station_id', 'end_station_name',
       'end_station_latitude', 'end_station_longitude', 'bike_id', 'user_type',
       'member_birth_year', 'member_gender', 'bike_share_for_all_trip'],
      dtype='object')

**Structure of the Dataset**  
There are 1863721 rows and 16 variables in the dataset(duration_sec, start_time, end_time, start_station_id, start_station_name, start_station_latitude, start_station_longitude, end_station_id, end_station_name, end_station_latitude, end_station_longitude, bike_id, user_type, member_birth_year, member_gender, bike_share_for_all_trip), 9 of these are numerical variables.

**Main features of interest in the dataset**  
We are more interested in figuring out what factors or features affect the place a trip is taken and the duration of the trips.

In [ ]:
**Features That will help support our investigation into our features of interest**  
Probably during the summer and holidays there will be more trip and for more time so the variables start_time and
end_time will probably have the major effect on the duration, we can also expect that subscriber takes more trips
than customers 
